app/src/main/java/io/apicurio/registry/services/auth/CustomAuthenticationMechanism.java

```diff
@@ -53,40 +73,102 @@
-                return oidcAuthenticationMechanism.authenticate(context, identityProviderManager);
-            } else {
-                final Pair<String, String> credentialsFromContext = CredentialsHelper.extractCredentialsFromContext(context);
-                if (credentialsFromContext != null) {
-                    OidcAuth oidcAuth = new OidcAuth(authServerUrl, clientId, clientSecret.get(), Optional.empty());
+            setAuditLogger(context);
+            if (fakeBasicAuthEnabled) {
+                final Pair<String, String> clientCredentials = CredentialsHelper.extractCredentialsFromContext(context);
+                if (null != clientCredentials) {
                     try {
-                        String jwtToken = oidcAuth.obtainAccessTokenWithBasicCredentials(credentialsFromContext.getLeft(), credentialsFromContext.getRight());
-                        if (jwtToken != null) {
-                            //If we manage to get a token from basic credentials, try to authenticate it using the fetched token using the identity provider manager
-                            return identityProviderManager
-                                    .authenticate(new TokenAuthenticationRequest(new AccessTokenCredential(jwtToken, context)));
-                        }
-                    } finally {
-                        oidcAuth.close();
+                        return authenticateWithClientCredentials(clientCredentials, context, identityProviderManager);
+                    } catch (NotAuthorizedException ex) {
+                        //Ignore exception, wrong credentials passed
+                        throw new AuthenticationFailedException();
                     }
                 } else {
-                    //If we cannot get a token, then try to authenticate using oidc provider as last resource
-                    return oidcAuthenticationMechanism.authenticate(context, identityProviderManager);
+                    return customAuthentication(context, identityProviderManager);
                 }
+            } else {
+                return customAuthentication(context, identityProviderManager);
+            }
+        } else {
+            return Uni.createFrom().nullItem();
+        }
+    }
+
+    public Uni<SecurityIdentity> customAuthentication(RoutingContext context, IdentityProviderManager identityProviderManager) {
+        if (clientSecret.isEmpty()) {
+            //if no secret is present, try to authenticate with oidc provider
+            return oidcAuthenticationMechanism.authenticate(context, identityProviderManager);
+        } else {
+            final Pair<String, String> credentialsFromContext = CredentialsHelper.extractCredentialsFromContext(context);
+            if (credentialsFromContext != null) {
+                String jwtToken = new OidcAuth(httpClient, clientId, clientSecret.get()).obtainAccessTokenPaswordGrant(credentialsFromContext.getLeft(), credentialsFromContext.getRight());
```

Comment:
caching of this token happens inside OidcAuth class, so if we don't save all the instances of this class that we create we won't be caching the token.

Reply:
This was not cached before and therefore is not cached now. In order to catch this, we need to store the pair clientId/secret (hashed), not the actual OidcAuth instance. IMO, that's is a separate task.

Reply:
but OidcAuth class caches internally the token and checks for expiration. anyway I agree that's a separate task

Reply:
+1